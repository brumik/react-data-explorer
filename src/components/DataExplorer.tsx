import '@patternfly/react-core/dist/styles/base.css';
import React, {
    FunctionComponent,
    useEffect,
    useState
} from 'react';
import {
    Button,
    Card,
    CardBody,
    CardTitle
} from '@patternfly/react-core';
import {
    FormApiProps,
    ChartFunctions,
    ChartSchemaElement,
    ChartEditor,
    ChartRenderer
} from '../';

interface Props {
    apis: FormApiProps[],
    schema: ChartSchemaElement[],
    functions: ChartFunctions,
    onSchemaChange?: (schema: ChartSchemaElement[]) => void
}

const DataExplorer: FunctionComponent<Props> = ({
    apis,
    schema: defaultSchema,
    functions,
    onSchemaChange = (d) => {
        /* eslint-disable-next-line */
        console.log(JSON.stringify(d));
    }
}) => {
    const [editorOpen, setEditorOpen] = useState(false);
    const [schema, setSchema] = useState(defaultSchema);
    const [temp, setTemp] = useState([] as ChartSchemaElement[]);

    useEffect(() => {
        if (!editorOpen) {
            setSchema([
                ...schema,
                ...temp
            ] as ChartSchemaElement[]);
            setTemp([]);
        }
    }, [editorOpen]);

    useEffect(() => {
        onSchemaChange(schema);
    }, [schema])

    return (
        <Card>
            <CardTitle>Data Explorer</CardTitle>
            <CardBody>
                <Button onClick={() => {
                    setEditorOpen(!editorOpen)
                }}>
                    {editorOpen && <span>Save Chart</span>}
                    {!editorOpen && <span>Add new chart</span>}
                </Button>
                { editorOpen && (
                    <ChartEditor
                        schema={schema}
                        apis={apis}
                        onSchemaChange={(updated) => setTemp(updated)}
                    />
                )}
                { !editorOpen && (
                    <ChartRenderer
                        data={{
                            charts: schema,
                            functions
                        }}
                    />
                )}
            </CardBody>
        </Card>
    );
};

export default DataExplorer;
