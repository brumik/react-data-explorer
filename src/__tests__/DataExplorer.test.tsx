import React from 'react'
import fetchMock from 'fetch-mock-jest';
import { act, render } from '@testing-library/react';
import DataExplorer from '../components/DataExplorer';
import {
    ChartSchemaElement,
    ChartKind,
    ChartLegendOrientation,
    ChartLegendPosition,
    ChartTopLevelType,
    functions
} from '../index';

const pieChartResponse = {
    "items": [
        { "host_count": 4655, "total_count": 578, "id": 2, "name": "organization_0" },
        { "host_count": 96225, "total_count": 456, "id": -2, "name": "" },
        { "host_count": 2995, "total_count": 411, "id": 4, "name": "organization_3" },
        { "host_count": 2975, "total_count": 410, "id": 3, "name": "organization_1" },
        { "host_count": 2975, "total_count": 410, "id": 1, "name": "organization_2" },
        { "host_count": 70720, "id": -1, "name": "26 Others" }
    ]
};

const pieChartSchema: ChartSchemaElement[] = [
    {
        id: 4000,
        kind: ChartKind.wrapper,
        type: ChartTopLevelType.pie,
        parent: null,
        props: {
            height: 300,
            x: '',
            y: 'host_count'
        },
        api: {
            params: {
                group_by: 'org',
                include_others: true,
                attributes: ['host_count'],
                sort_by: `total_count:desc`
            },
            url: 'http://piechart.url',
            method: 'GET'
        },
        legend: {
            interactive: true,
            orientation: ChartLegendOrientation.vertical,
            position: ChartLegendPosition.right
        }
    }
];

describe('Data Explorer', () => {
    test('should render empty DataExplorer', () => {
        const { container } = render(<DataExplorer
            apis={[]}
            schema={[]}
            onSchemaChange={() => ({})}
            functions={functions}
        />)
        expect(container).toMatchSnapshot()
    });

    test('should render pie chart DataExplorer', async () => {
        fetchMock.get({ url: 'http://piechart.url' }, pieChartResponse);
        let container = null;
        await act(async () => {
            container = render(<DataExplorer
                apis={[]}
                schema={pieChartSchema}
                onSchemaChange={() => ({})}
                functions={functions}
            />).container;
        });

        expect(container).toMatchSnapshot()
    });
});
