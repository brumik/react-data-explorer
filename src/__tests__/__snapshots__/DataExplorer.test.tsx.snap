// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Explorer should render empty DataExplorer 1`] = `
<div>
  <article
    class="pf-c-card"
    data-ouia-component-id="OUIA-Generated-Card-1"
    data-ouia-component-type="PF4/Card"
    data-ouia-safe="true"
    id=""
  >
    <div
      class="pf-c-card__title"
    >
      Data Explorer
    </div>
    <div
      class="pf-c-card__body"
    >
      <button
        aria-disabled="false"
        class="pf-c-button pf-m-primary"
        data-ouia-component-id="OUIA-Generated-Button-primary-1"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe="true"
        type="button"
      >
        <span>
          Add new chart
        </span>
      </button>
    </div>
  </article>
</div>
`;

exports[`Data Explorer should render pie chart DataExplorer 1`] = `
<div>
  <article
    class="pf-c-card"
    data-ouia-component-id="OUIA-Generated-Card-2"
    data-ouia-component-type="PF4/Card"
    data-ouia-safe="true"
    id=""
  >
    <div
      class="pf-c-card__title"
    >
      Data Explorer
    </div>
    <div
      class="pf-c-card__body"
    >
      <button
        aria-disabled="false"
        class="pf-c-button pf-m-primary"
        data-ouia-component-id="OUIA-Generated-Button-primary-2"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe="true"
        type="button"
      >
        <span>
          Add new chart
        </span>
      </button>
      <div>
        <div
          style="height: 300px;"
        >
          <div
            class="pf-c-chart"
            style="pointer-events: none; position: relative; width: 100%; height: 100%;"
          >
            <svg
              height="300"
              role="img"
              style="pointer-events: all; width: 100%; height: 100%;"
              viewBox="0 0 0 300"
              width="0"
            >
              <g>
                <path
                  d="M0,0Z"
                  role="presentation"
                  shape-rendering="auto"
                  style="fill: #06c; padding: 8px; stroke: transparent; stroke-width: 1;"
                  transform="translate(-40, 140)"
                />
                <path
                  d="M0,0Z"
                  role="presentation"
                  shape-rendering="auto"
                  style="fill: #8bc1f7; padding: 8px; stroke: transparent; stroke-width: 1;"
                  transform="translate(-40, 140)"
                />
                <path
                  d="M0,0Z"
                  role="presentation"
                  shape-rendering="auto"
                  style="fill: #002f5d; padding: 8px; stroke: transparent; stroke-width: 1;"
                  transform="translate(-40, 140)"
                />
                <path
                  d="M0,0Z"
                  role="presentation"
                  shape-rendering="auto"
                  style="fill: #519de9; padding: 8px; stroke: transparent; stroke-width: 1;"
                  transform="translate(-40, 140)"
                />
                <path
                  d="M0,0Z"
                  role="presentation"
                  shape-rendering="auto"
                  style="fill: #004b95; padding: 8px; stroke: transparent; stroke-width: 1;"
                  transform="translate(-40, 140)"
                />
                <path
                  d="M0,0Z"
                  role="presentation"
                  shape-rendering="auto"
                  style="fill: #06c; padding: 8px; stroke: transparent; stroke-width: 1;"
                  transform="translate(-40, 140)"
                />
              </g>
              <g>
                <rect
                  height="185.42999999999998"
                  role="presentation"
                  shape-rendering="auto"
                  style="fill: none;"
                  vector-effect="non-scaling-stroke"
                  width="132.896875"
                  x="0"
                  y="49.28500000000001"
                />
                <path
                  d="M 9.128, 68.15700000000001
      h9.744
      v-9.744
      h-9.744
      z"
                  role="presentation"
                  shape-rendering="auto"
                  style="fill: #06c;"
                />
                <path
                  d="M 9.128, 99.06200000000001
      h9.744
      v-9.744
      h-9.744
      z"
                  role="presentation"
                  shape-rendering="auto"
                  style="fill: #8bc1f7;"
                />
                <path
                  d="M 9.128, 129.96699999999998
      h9.744
      v-9.744
      h-9.744
      z"
                  role="presentation"
                  shape-rendering="auto"
                  style="fill: #002f5d;"
                />
                <path
                  d="M 9.128, 160.872
      h9.744
      v-9.744
      h-9.744
      z"
                  role="presentation"
                  shape-rendering="auto"
                  style="fill: #519de9;"
                />
                <path
                  d="M 9.128, 191.777
      h9.744
      v-9.744
      h-9.744
      z"
                  role="presentation"
                  shape-rendering="auto"
                  style="fill: #004b95;"
                />
                <path
                  d="M 9.128, 222.68200000000002
      h9.744
      v-9.744
      h-9.744
      z"
                  role="presentation"
                  shape-rendering="auto"
                  style="fill: #06c;"
                />
                <text
                  direction="inherit"
                  dx="0"
                  id="legend-4000-labels-0"
                  x="30.8"
                  y="68.25500000000001"
                >
                  <tspan
                    dx="0"
                    dy="0"
                    style="cursor: pointer; font-family: \\"RedHatText\\",\\"Overpass\\",overpass,helvetica,arial,sans-serif; font-size: 14px; letter-spacing: normal; padding: 10px; stroke: transparent; fill: #252525;"
                    text-anchor="start"
                    x="30.8"
                  >
                    organization_0
                  </tspan>
                </text>
                <text
                  direction="inherit"
                  dx="0"
                  id="legend-4000-labels-1"
                  x="30.8"
                  y="99.16000000000001"
                >
                  <tspan
                    dx="0"
                    dy="0"
                    style="cursor: pointer; font-family: \\"RedHatText\\",\\"Overpass\\",overpass,helvetica,arial,sans-serif; font-size: 14px; letter-spacing: normal; padding: 10px; stroke: transparent; fill: #252525;"
                    text-anchor="start"
                    x="30.8"
                  >
                    No Name
                  </tspan>
                </text>
                <text
                  direction="inherit"
                  dx="0"
                  id="legend-4000-labels-2"
                  x="30.8"
                  y="130.065"
                >
                  <tspan
                    dx="0"
                    dy="0"
                    style="cursor: pointer; font-family: \\"RedHatText\\",\\"Overpass\\",overpass,helvetica,arial,sans-serif; font-size: 14px; letter-spacing: normal; padding: 10px; stroke: transparent; fill: #252525;"
                    text-anchor="start"
                    x="30.8"
                  >
                    organization_3
                  </tspan>
                </text>
                <text
                  direction="inherit"
                  dx="0"
                  id="legend-4000-labels-3"
                  x="30.8"
                  y="160.97"
                >
                  <tspan
                    dx="0"
                    dy="0"
                    style="cursor: pointer; font-family: \\"RedHatText\\",\\"Overpass\\",overpass,helvetica,arial,sans-serif; font-size: 14px; letter-spacing: normal; padding: 10px; stroke: transparent; fill: #252525;"
                    text-anchor="start"
                    x="30.8"
                  >
                    organization_1
                  </tspan>
                </text>
                <text
                  direction="inherit"
                  dx="0"
                  id="legend-4000-labels-4"
                  x="30.8"
                  y="191.875"
                >
                  <tspan
                    dx="0"
                    dy="0"
                    style="cursor: pointer; font-family: \\"RedHatText\\",\\"Overpass\\",overpass,helvetica,arial,sans-serif; font-size: 14px; letter-spacing: normal; padding: 10px; stroke: transparent; fill: #252525;"
                    text-anchor="start"
                    x="30.8"
                  >
                    organization_2
                  </tspan>
                </text>
                <text
                  direction="inherit"
                  dx="0"
                  id="legend-4000-labels-5"
                  x="30.8"
                  y="222.78"
                >
                  <tspan
                    dx="0"
                    dy="0"
                    style="cursor: pointer; font-family: \\"RedHatText\\",\\"Overpass\\",overpass,helvetica,arial,sans-serif; font-size: 14px; letter-spacing: normal; padding: 10px; stroke: transparent; fill: #252525;"
                    text-anchor="start"
                    x="30.8"
                  >
                    26 Others
                  </tspan>
                </text>
              </g>
            </svg>
            <div
              style="z-index: 99; position: absolute; top: 0px; left: 0px; width: 100%; height: 100%;"
            >
              <svg
                height="300"
                style="overflow: visible; width: 100%; height: 100%;"
                viewBox="0 0 0 300"
                width="0"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </article>
</div>
`;
